File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        
        def dfs(node, maxVal):

            # Base Case
            if not node:
                return 0

            # Recursive Case
            goodnode = 1 if node.val >= maxVal else 0
            maxVal = max(node.val, maxVal)

            leftSum  = dfs(node.left, maxVal)
            rightSum = dfs(node.right, maxVal)

            res = goodnode + leftSum + rightSum
            return res 


        return dfs(root, root.val)

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-4.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {

        if ( root == null ) return null;
        TreeNode curr = root;

        while ( curr != null ) {

            if ( curr.val > p.val && curr.val > q.val) {
                curr = curr.left;
            } else if ( curr.val < p.val && curr.val < q.val) {
                curr = curr.right;
            } else {
                return curr;
            }
        }
        return null;
    }
}



// class Solution {
//     public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {

//         if ( root == null ) return null;

//         if ( root.val > p.val && root.val > q.val) {
//             return lowestCommonAncestor(root.left, p, q);
//         }

//         if ( root.val < p.val && root.val < q.val) {
//             return lowestCommonAncestor(root.right, p, q);
//         }

//         return root; 
 
//     }
// }
